pygem/setup/config.py

```diff
@@ -7,42 +7,336 @@
-
-def ensure_config(overwrite=False):
-    isfile = os.path.isfile(config_file)
-    if isfile and overwrite:
-        overwrite = None
-        while overwrite is None:
-            # Ask the user for a y/n response
-            response = input(f"PyGEM configuration.yaml file already exist ({config_file}), do you wish to overwrite (y/n):").strip().lower()
-            # Check if the response is valid
-            if response == 'y' or response == 'yes':
-                overwrite = True
-            elif response == 'n' or response == 'no':
-                overwrite = False
-            else:
-                print("Invalid input. Please enter 'y' or 'n'.")
+class ConfigManager:
+    def __init__(self, config_filename='config.yaml', base_dir=None, overwrite=False):
+        """initialize the ConfigManager class"""
+        self.config_filename = config_filename
+        self.base_dir = base_dir or os.path.join(os.path.expanduser('~'), 'PyGEM')
+        self.config_path = os.path.join(self.base_dir, self.config_filename)
+        self.source_config_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "config.yaml")
+        self.overwrite = overwrite
+        self.ensure_config()
     
-    if (not isfile) or (overwrite):
-        os.makedirs(basedir, exist_ok=True)  # Ensure the base directory exists
-        shutil.copy(source_config_file, config_file)  # Copy the file
-        print(f"Copied default configuration to {config_file}")
+    def ensure_config(self):
+        """Ensure the configuration file exists, creating or overwriting it if necessary"""
+        if not os.path.isfile(self.config_path) or self.overwrite:
+            self.copy_source_config()
+
+    def copy_source_config(self):
+        """Copy the default configuration file to the expected location"""
+
+        os.makedirs(self.base_dir, exist_ok=True)
+        shutil.copy(self.source_config_path, self.config_path)
+        print(f"Copied default configuration to {self.config_path}")
+        
+    def read_config(self, validate=True):
+        """Read the configuration file and return its contents as a dictionary while preserving formatting."""
+        ryaml = ruamel.yaml.YAML()
+        with open(self.config_path, 'r') as f:
+            user_config = ryaml.load(f)  # Using ruamel.yaml for preservation
+
+        if validate:
+            self.validate_config(user_config)
+
+        return user_config
+
+    def write_config(self, config):
+        """Write the configuration dictionary to the file while preserving quotes."""
+        ryaml = ruamel.yaml.YAML()
+        ryaml.preserve_quotes = True
+        with open(self.config_path, 'w') as file:
+            ryaml.dump(config, file)  # This will preserve quotes
+    
+    def update_config(self, updates):
+        """Update multiple keys in the YAML configuration file while preserving quotes and original types."""
+        config = self.read_config(validate=False)  # Read existing config
```

Comment:
Unnecessary comment (it's clear from the name of the function `read_config`)
```suggestion
        config = self.read_config(validate=False)
```